statshouse.hpp

```diff
@@ -695,6 +708,473 @@ class TransportUDP : public TransportUDPBase {
+		}
+	private:
+		mutable Registry *registry;
+		TransportUDP::MetricBuilder key;
+	};
+	MetricBuilder metric(string_view name) {
+		return {this, name};
+	}
+	// If called once, then you need to call it until the end of the life of the registry.
+	void set_external_time(uint32_t timestamp) {
+		time_external = timestamp;
+	}
+	// Call once during initialization, if set then you are responsible for calling "flush".
+	void disable_incremental_flush() {
+		incremental_flush_disabled = true;
+	}
+	void flush() {
+		if (incremental_flush_disabled) {
+			flush(time_now() - 1);
+		} else {
+			flush(time_now() - 2);
+		}
+	}
+	void set_max_udp_packet_size(size_t n) {
+		std::lock_guard<std::mutex> transport_lock{transport_mu};
+		transport.set_max_udp_packet_size(n);
+	}
+	void set_max_bucket_size(size_t n) {
+		max_bucket_size.store(n, std::memory_order_relaxed);
+	}
+	void set_default_env(const std::string &env) {
+		std::lock_guard<std::mutex> transport_lock{transport_mu};
+		transport.set_default_env(env);
+	}
+	class Clock {
+	public:
+		explicit Clock(Registry *registry)
+			: thread{start(registry)} {
+		}
+		~Clock() {
+			stop = true;
+			thread.join();
+		}
+	private:
+		std::thread start(Registry *registry) {
+			registry->set_external_time(std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now()).time_since_epoch().count());
+			return std::thread(&Clock::run, this, registry);
+		}
+		void run(Registry *registry) {
+			for (uint32_t prev = 0; !stop;) {
+				auto atime = std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now()) + std::chrono::seconds{1};
+				std::this_thread::sleep_until(atime);
+				if (stop) {
+					break;
+				}
+				registry->set_external_time(atime.time_since_epoch().count());
+			}
+		}
+		std::thread thread;
+		std::atomic_bool stop{};
+	};
+private:
+	bool update_multivalue_by_key(const TransportUDP::MetricBuilder &key, multivalue_view &value) {
+		auto timestamp = time_now();
+		return update_multivalue_by_ref(get_or_create_bucket(key, timestamp), timestamp, value);
+	}
+	bool update_multivalue_by_ref(const std::shared_ptr<bucket> &ptr, multivalue_view &value) {
+		return update_multivalue_by_ref(ptr, time_now(), value);
+	}
+	bool update_multivalue_by_ref(const std::shared_ptr<bucket> &ptr, uint32_t timestamp, multivalue_view &value) {
+		auto limit = max_bucket_size.load(std::memory_order_relaxed);
+		{
+			std::lock_guard<std::mutex> bucket_lock{ptr->mu};
+			if (timestamp <= ptr->timestamp) {
+				ptr->value.write(value, limit);
+				if (value.empty()) {
+					return true; // fast path
+				}
+				// protect against clock going backwards before falling through
+				timestamp = ptr->timestamp;
+			} else {
+				// next second
+			}
+		}
+		// split
+		size_t queue_size = 0;
+		{
+			std::lock_guard<std::mutex> lock{mu};
+			queue.emplace_back(std::move(*ptr->queue_ptr));
+			*ptr->queue_ptr = alloc_bucket(ptr->key, ptr->timestamp);
```

Comment:
Possible use-after-free leading to undefined behaviour.

Honestly, the whole `queue_ptr` structure and code using it is too obscure and overall dangerous. Could you please describe what exactly is happening there and why do you need to remember raw pointers???

Reply:
The structure is quite complex, i agree. I don't see use-after-free here though. Could you elaborate on this?